code/espurna/sensors/ADE7953Sensor.h

```diff
@@ -0,0 +1,223 @@
+            if (current_rms1 < 2000) {             // No load threshold (20mA)
+                current_rms1 = 0;
+                active_power1 = 0;
+            } else {
+                active_power1 = (int32_t)Ade7953Read(0x313) * -1;  // Relay 1
+                active_power1 = (active_power1 > 0) ? active_power1 : 0;
+            }
+            current_rms2 = Ade7953Read(0x31A);     // Relay 2
+            if (current_rms2 < 2000) {             // No load threshold (20mA)
+                current_rms2 = 0;
+                active_power2 = 0;
+            } else {
+                active_power2 = (int32_t)Ade7953Read(0x312);  // Relay 2
+                active_power2 = (active_power2 > 0) ? active_power2 : 0;
+            }
+            _voltage = (float) voltage_rms / ADE7953_UREF;            
+            
+            writeFloat(
+                ADE7953_ALL_RELAYS, 
+                (float)(current_rms1 + current_rms2) / (ADE7953_IREF * 10), 
+                (float)(active_power1 + active_power2) / (ADE7953_PREF / 10)
+            );
+            writeFloat(
+                ADE7953_RELAY_1, 
+                (float) current_rms1 / (ADE7953_IREF * 10), 
+                (float) active_power1 / (ADE7953_PREF / 10)
+            );    
+            writeFloat(
+                ADE7953_RELAY_2, 
+                (float)current_rms2 / (ADE7953_IREF * 10), 
+                (float)active_power2 / (ADE7953_PREF / 10)
+            );
+        }
+
+        void writeFloat(unsigned int relay, float current, float power) {
+            float* current_reading = &_readings[relay].current;
+            float* power_reading = &_readings[relay].power;
+            *current_reading = current;
+            *power_reading = power;
+        }
+
+        // Current value for slot # index
+        double value(unsigned char index) {
+            int relay = index / ADE7953_TOTAL_DEVICES;
+            index = index - (relay * ADE7953_TOTAL_DEVICES);
+            if (index == 0) return _voltage;
+            if (index == 1) return _readings[relay].current;
+            if (index == 2) return _readings[relay].power;     
+            if (index == 3) return _readings[relay].energy;
+            return 0;
+        }
+
+    protected:  
+        void _init() {                     
+            delay(100);                     // Need 100mS to init ADE7953
+            Ade7953Write(0x102, 0x0004);    // Locking the communication interface (Clear bit COMM_LOCK), Enable HPF
+            Ade7953Write(0x0FE, 0x00AD);    // Unlock register 0x120
+            Ade7953Write(0x120, 0x0030);    // Configure optimum setting       
+        
+            for(unsigned int i = 0; i > ADE7953_TOTAL_DEVICES ; i++) {
+                reading_t reading;
+                reading.current = 0.0;                
+                reading.power = 0.0;
+                reading.energy = 0.0;
+                _readings.push_back(reading);
```

Comment:
`_readings.resize(ADE7953_TOTAL_DEVICES);` should do the same thing.
i have vague understanding of what default struct values are in this case, but you can just bind these down below where `struct reading_t { ... };` is.